src/util.c

```diff
@@ -2119,28 +2112,100 @@ ni_parse_hex(const char *string, unsigned char *data, unsigned int datasize)
 	return len;
 }
 
-const char *
-ni_format_bitmap(ni_stringbuf_t *buf, const ni_intmap_t *map,
-		unsigned int flags, const char *sep)
+unsigned int
+ni_parse_bitmap_array(unsigned int *mask, const ni_intmap_t *map, const ni_string_array_t *array,
+			ni_string_array_t *invalid)
 {
 	unsigned int i, flag;
+	const char *name;
+	unsigned int err = 0;
+
+	if (!mask || !map || !array)
+		return -1U;
+
+	for (i = 0; i < array->count; ++i) {
+		name = array->data[i];
+
+		if (!ni_parse_uint_mapped(name, map, &flag) &&
+				flag < 8 * sizeof(*mask)) {
+			*mask |= NI_BIT(flag);
+		} else {
+			if (invalid)
+				ni_string_array_append(invalid, name);
+			err++;
+		}
+	}
+
+	return err;
+}
+
+unsigned int
+ni_parse_bitmap_string(unsigned int *mask, const ni_intmap_t *map, const char *input,
+		const char *sep, ni_string_array_t *invalid)
+{
+	ni_string_array_t array = NI_STRING_ARRAY_INIT;
+	unsigned int err;
+
+	if (!mask || !map || !input)
+		return -1U;
+
+	ni_string_split(&array, input, sep, 0);
+	err = ni_parse_bitmap_array(mask, map, &array, invalid);
+	ni_string_array_destroy(&array);
+
+	return err;
+}
+
+unsigned int
+ni_format_bitmap_array(ni_string_array_t *array, const ni_intmap_t *map,
+		unsigned int mask, unsigned int *done)
+{
+	unsigned int flag;
+
+	if (!array || !map)
+		return -1U;
+
+	for (; map->name; ++map) {
+		flag = NI_BIT(map->value);
+		if (mask & flag) {
+			if (ni_string_array_append(array, map->name) < 0)
+				continue;
+
+			mask &= ~flag;
+			if (done)
+				*done |= flag;
+		}
+	}
+
+	return mask;
```

Comment:
This also differs from your version  https://github.com/mtomaschewski/wicked/blob/bitmap-utils/src/util.c#L2084

But I think in your version it was broken and actually returned if one flag of `map` wasn't set at all.
The other possibility is returning 0 here as well and just let the callee get the wrong bits with
```
unknown_bit_s = done ^ mask;

Reply:
Yes, it was broken an counting valid but unset bits as errors -- yours is fine and returns the unknown bits as "error".
The format_string is less strict and returns NULL only on invalid args (-1U) and not all errors (unknown bits), but this is probably OK to format what we can/understand and discard any unknown/unnamed bits.